yajsapi/executor/agreements_pool.ts

```diff
@@ -0,0 +1,192 @@
+    buffered_agreement.worker_task = task;
+  }
+  private async _get_agreement(): Promise<[Agreement, NodeInfo] | undefined> {
+    const emit = this.emitter;
+    const available_agreements =
+      [...this._agreements.values()].filter(agr => agr.worker_task === undefined);
+    if (available_agreements.length > 0) {
+      const buffered_agreement
+        = available_agreements[Math.floor(Math.random() * available_agreements.length)];
+      logger.debug(`Reusing agreement. id: ${buffered_agreement.agreement.id()}`);
+      return [buffered_agreement.agreement, buffered_agreement.node_info];
+    }
+    if (this._offer_buffer.size === 0) { return; }
+    let _sample =
+      [...this._offer_buffer.entries()]
+      .map(x => { return { obj: x, rnd: Math.random() }; })
+      .sort((a, b) => a.rnd - b.rnd)
+      .map(x => x.obj)
+      .reduce((acc, item) => item[1].score > acc[1].score ? item : acc);
+    let [provider_id, offer] = _sample;
+    this._offer_buffer.delete(provider_id);
+    if (this._cancellation_token.cancelled) return;
+    try {
+      const agreement = await offer.proposal.create_agreement();
+      try {
+        const agreement_details = await agreement.details()
+        const provider_activity = <Activity>agreement_details.provider_view().extract(new Activity());
+        const requestor_activity = <Activity>agreement_details.requestor_view().extract(new Activity());
+        const node_info = <NodeInfo>agreement_details.provider_view().extract(new NodeInfo());
+        logger.debug(`New agreement. id: ${agreement.id()}, provider: ${node_info.name.value}`);
+        emit(
+          new events.AgreementCreated({
+            agr_id: agreement.id(),
+            provider_id: provider_id,
+            provider_info: node_info,
+          })
+        );
+        if (this._cancellation_token.cancelled) return;
+        if (!(await agreement.confirm())) {
+          emit(new events.AgreementRejected({ agr_id: agreement.id() }));
+          this._rejecting_providers.add(provider_id);
+          return;
+        }
+        this._rejecting_providers.delete(provider_id);
+        this._agreements.set(
+          agreement.id(),
+          new BufferedAgreement(
+            agreement,
+            node_info,
+            undefined,
+            provider_activity.multi_activity.value && requestor_activity.multi_activity.value
+          )
+        )
+        emit(new events.AgreementConfirmed({ agr_id: agreement.id() }));
+        this._confirmed += 1;
+        return [agreement, node_info];
+      } catch (error) {
+        logger.debug(`Cannot get agreement details. id: ${agreement.id()}`);
+        emit(new events.AgreementRejected({ agr_id: agreement.id() }));
+        return;
+      }
+    } catch (e) {
+      emit(
+        new events.ProposalFailed({
+          prop_id: offer.proposal.id(),
+          reason: e.toString(),
+        })
+      );
+      throw e;
+    }
```

Comment:
I'd consider rewriting this:
```javascript
try {
   const agreement = await offer.proposal.create_agreement();
   try { 
      <code>
   } catch (error) {
      ... return;
   }
} catch (e) {
   ... throw e;
}
```
to:
```javascript
try {
   const agreement = await offer.proposal.create_agreement();
} catch (e) {
   ... throw e;
}
try {
   <code>
} catch (error) {
   ... return;
}
```
to avoid nesting `try ... catch` statements.

Reply:
But then `agreement` is not visible in `<code>`, so it's necessary to use `let` instead of `const` and put it above the first `try`.